File: ml/train_two_tower.py
import tensorflow as tf
import json
import os

from two_tower import TwoTowerEmbeddings
from dataset.build_dataset import build_dataset
from dataset.build_user_tensor import build_user_tensor
from api.get_data import load_dataset


def train():

    try:
        users, songs, sequences = load_dataset()

        all_songs = songs

        if len(all_songs) == 0:
            raise Exception("Songs not found")

        embedding_length = len(all_songs[0]["embedding"])

        user_inputs = []
        song_inputs = []
        labels = []

        for user_id in users:

            sequence = sequences.get(str(user_id), [])

            samples = build_dataset(
                sequence=sequence,
                all_songs=all_songs,
                sequence_length=10,
                negative_per_positive=1,
                embedding_length=embedding_length
            )

            for s in samples:

                user_inputs.append(s["user_tensor"])

                song_inputs.append(
                    s["song_embedding"]
                )

                labels.append([
                    float(s["is_positive"])
                ])

        if len(user_inputs) == 0:
            raise Exception("No training samples")

        userX = tf.convert_to_tensor(
            user_inputs,
            dtype=tf.float32
        )

        songX = tf.convert_to_tensor(
            song_inputs,
            dtype=tf.float32
        )

        y = tf.convert_to_tensor(
            labels,
            dtype=tf.float32
        )

        print("DATA INFO", {
            "users": len(users),
            "tracks": len(all_songs),
            "samples": len(user_inputs)
        })

        base_model = TwoTowerEmbeddings(
            sequence_length=10,
            embedding_length=embedding_length
        )

        user_input = tf.keras.Input(
            shape=(10, embedding_length + 3)
        )

        song_input = tf.keras.Input(
            shape=(embedding_length,)
        )

        user_vec = base_model.user_encoder(user_input)
        song_vec = base_model.song_encoder(song_input)

        dot = tf.keras.layers.Dot(
            axes=1
        )([user_vec, song_vec])

        output = tf.keras.layers.Activation(
            "sigmoid"
        )(dot)

        training_model = tf.keras.Model(
            inputs=[user_input, song_input],
            outputs=output
        )

        training_model.compile(
            optimizer=tf.keras.optimizers.Adam(0.001),
            loss=tf.keras.losses.BinaryCrossentropy(),
            metrics=["accuracy"]
        )

        print("TRAINING STARTED")

        training_model.fit(
            [userX, songX],
            y,
            epochs=5,
            batch_size=64,
            shuffle=True
        )

        print("TRAINING FINISHED")

        os.makedirs("./data", exist_ok=True)
        
        training_model.save("./data/two_tower_model.keras")

        user_embeddings = []
        song_embeddings = []

        for user_id in users:

            sequence = sequences.get(str(user_id), [])

            user_tensor = tf.convert_to_tensor(
                [build_user_tensor(sequence, embedding_length)],
                dtype=tf.float32
            )

            emb = base_model.user_encoder(
                user_tensor
            ).numpy()[0]

            user_embeddings.append({
                "userId": user_id,
                "embedding": emb.tolist()
            })

        for song in all_songs:

            song_tensor = tf.convert_to_tensor(
                [song["embedding"]],
                dtype=tf.float32
            )

            emb = base_model.song_encoder(
                song_tensor
            ).numpy()[0]

            song_embeddings.append({
                "songId": song["id"],
                "embedding": emb.tolist()
            })

        with open("./data/user_embeddings.json", "w") as f:
            json.dump(user_embeddings, f)

        with open("./data/song_embeddings.json", "w") as f:
            json.dump(song_embeddings, f)

    except Exception as err:
        print("TRAIN FAILED")
        print(err)


train()

File: server/services/recommendations/trainTwoTower.js
import 'dotenv/config'
import tf from '@tensorflow/tfjs'
import sequelize from '../../db.js'

import UserDataService from './userDataService.js'
import TwoTowerEmbeddings from './TwoTowerEmbeddings.js'
import { buildDataset } from './dataset/buildDataset.js'

const userDataService = new UserDataService()

export async function train() {
  try {
    await sequelize.authenticate()

    const embeddingMap =
      await userDataService.getAllSongsEmbeddingMap()

    const allSongs = Array.from(embeddingMap.entries()).map(
      ([id, embedding]) => ({ id, embedding })
    )

    const users = await userDataService.getAllUsers()

    const embeddingLength = allSongs[0].embedding.length

    const userInputs = []
    const songInputs = []
    const labels = []

    for (const userId of users) {
      const sequence =
        await userDataService.getUserSequence(userId)

      const samples =
        buildDataset(sequence, allSongs, 10, 1, embeddingLength)

      for (const s of samples) {
        userInputs.push(s.userTensor)
        songInputs.push(s.songEmbedding)
        labels.push([s.isPositive])
      }
    }

    if (!userInputs.length) {
      throw new Error('No training samples')
    }

    const userX = tf.tensor3d(
      userInputs,
      [userInputs.length, 10, embeddingLength + 3]
    )

    const songX = tf.tensor2d(songInputs)
    const y = tf.tensor2d(labels)

    console.log('DATA INFO', {
      users: users.length,
      tracks: allSongs.length,
      samples: userInputs.length
    })

    const model = new TwoTowerEmbeddings({
      sequenceLength: 10,
      embeddingLength
    })

    const userInput = model.userEncoder.input
    const songInput = model.songEncoder.input

    const userVec = model.userEncoder.apply(userInput)
    const songVec = model.songEncoder.apply(songInput)

    const dot = tf.layers.dot({ axes: -1 }).apply([
      userVec,
      songVec
    ])

    const output = tf.layers.activation({
      activation: 'sigmoid'
    }).apply(dot)

    const trainingModel = tf.model({
      inputs: [userInput, songInput],
      outputs: output
    })

    trainingModel.compile({
      optimizer: tf.train.adam(0.001),
      loss: 'binaryCrossentropy',
      metrics: ['accuracy']
    })

    console.log('TRAINING STARTED')

    await trainingModel.fit(
      [userX, songX],
      y,
      {
        epochs: 5,
        batchSize: 64,
        shuffle: true,
        callbacks: {
          onEpochEnd: (epoch, logs) => {
            console.log({
              epoch: epoch + 1,
              loss: logs.loss,
              acc: logs.acc || logs.accuracy
            })
          }
        }
      }
    )

    console.log('TRAINING FINISHED')

    const userEmbeddings = await exportUserEmbeddings(
      model,
      users,
      embeddingLength
    )

    const songEmbeddings = await exportSongEmbeddings(
      model,
      allSongs
    )

    fs.writeFileSync('./data/user_embeddings.json', JSON.stringify(userEmbeddings))
    fs.writeFileSync('./data/song_embeddings.json', JSON.stringify(songEmbeddings))

  } catch (err) {
    console.error('TRAIN FAILED')
    console.error(err)
  } finally {
    await sequelize.close()
  }
}